Merge `release/4.0` into `main`

rozbb opened the pull request:
Also bumps version to `4.0.0-pre.3`.

This will say in the README that the current beta is `4.0.0-pre.3`, so we should probably `cargo publish` right after the merge.

isislovecruft committed:
Pin zeroize dependency to =1.3.0 to maintain MSRV 1.41.

isislovecruft committed:
Merge branch 'main' into develop

isislovecruft committed:
Merge branch 'develop' into release/4.0

PaulGrandperrin committed:
Update rand_core to 0.6

https://github.com/rust-random/rand/blob/master/rand_core/CHANGELOG.md
This new version makes using rand_core on wasm seamless (thanks to the update of getrandom to v0.2)
The crate compiles well with this PR, but since some `rand_core` traits are publicly exposed in this crate's API, this is strictly speaking a breaking change.

PaulGrandperrin committed:
Update rand to 0.8

isislovecruft committed:
Merge remote-tracking branch 'PaulGrandperrin/patch-1' into develop

isislovecruft committed:
Update CHANGELOG and README for 4.x alpha.

isislovecruft committed:
Revert "Maintain legacy 3.x support for lookup tables."

This reverts commit 0da8f08d6582cbc35af6ef1da0e7917d4f5ddfed.

isislovecruft committed:
Bump version to 4.0.0-pre.0.

isislovecruft committed:
Merge branch 'develop' into release/4.0

isislovecruft committed:
Merge branch 'develop' into release/4.0

isislovecruft committed:
Merge branch 'develop' into release/4.0

isislovecruft committed:
Bump alpha curve25519-dalek version to 4.0.0-pre.1.

isislovecruft committed:
Merge branch 'main' into develop

isislovecruft committed:
Merge branch 'develop' into release/4.0

isislovecruft committed:
Bump sha2 and digest dependencies to 0.10.

isislovecruft committed:
Merge branch 'feature/digest-sha2-v0.10' into release/4.0

isislovecruft committed:
Merge branch 'main' into develop

isislovecruft committed:
Merge branch 'develop' into release/4.0

isislovecruft committed:
Bump curve25519-dalek version to 4.0.0-pre.2.

rozbb committed:
Relax zeroize dependency and bump MSRV (#412)

rozbb committed:
Updated to edition 2021 (#413)

rozbb committed:
Noted edition change in changelog

k0nserv committed:
Remove byteorder (#418)

Instead of having a dependency on `byteorder`, use methods from the
standard library(`{to,from}_le_bytes`).

rozbb committed:
Fixed doc warnings

rozbb committed:
Wibble

rozbb committed:
Made docs makefile use nightly

rozbb committed:
Cherry-picked Github actions from 1f9e527

rozbb committed:
Fix AVX2 and AVX-512 builds (#419)

Build was broken by 5758b8c.
This adds a regression test to CI so it doesn't happen again

rozbb committed:
Undid the bad reformatting job from 5758b8c

rozbb committed:
Added #[rustfmt::skip] where necessary

rozbb committed:
Whitespace formatting fixes on the [rustfmt::skip] functions

rozbb committed:
Added more #[rusfmt::skip]

rozbb committed:
cargo fmt

rozbb committed:
Merge pull request #424 from rozbb/rustfmt

Did a full formatting pass on the crate. Read the comments in the referenced for
steps to verify that the PR made no semantic changes to the code. Huge credit
goes to @robjtede for this work.

pornin committed:
Fix double-and-compress on Ristretto identity (issue #398). (#399)

pinkforest committed:
Include README.md into the crate Documentation

As proposed in #426 this commit changes to include README.md into
the crate documentation instead of maintaining a carbon copy in
the code.

This helps to keep the documentation in a single place without
duplicating the documentation in multiple places leading to less
errors and out of date documentation.

rozbb committed:
Merge pull request #429 from pinkforest/docs/readme/de-dup

Use #[include_str] for top-level docs

pinkforest committed:
Bump criterion to 0.4.0 (#432)

tarcieri committed:
Implement simplified backend selection (#428)

As proposed in #414, this commit changes the backend selection approach,
introspecting `target_pointer_width` to select `u32_backend` vs
`u64_backend` (or `fiat_u32_backend`/`fiat_u64_backend` if the
`fiat_backend` feature is enabled).

This helps eliminate the use of non-additive features, and also the
rather confusing errors that happen if multiple backends are selected
(i.e. thousands of lines of rustc errors).

The selection logic checks if `target_pointer_width = "64"` and uses the
64-bit backend, or falls back to the 32-bit backend otherwise. This
means the crate will always have a valid backend regardless of the
pointer width, although there may be odd edge cases for exotic platforms
which would optimally use the 64-bit backend but have a non-"64" target
pointer width for whatever reason. We can handle those cases as they
come up.

tarcieri committed:
CI: fix `build-simd` job (#436)

The `RUSTFLAGS` were getting applied to build scripts, which caused them
to crash with SIGILL.

According to this issue, RUSTFLAGS won't be applied to build scripts
when cross-compiling by passing the `--target` attribute:

https://github.com/rust-lang/cargo/issues/4423

This attempts to work around the problem by explicitly passing:

    --target x86_64-unknown-linux-gnu

tarcieri committed:
Add alloc feature gates to simd tests that need it (#433)

tarcieri committed:
Use `include_str!` for `.md` inclusion in rustdoc (#434)

Previously a now-removed nightly-only feature was used, but now that
it's stable, `include_str!` can be used for all of these cases.

rozbb committed:
Deprecate `EdwardsPoint::hash_from_bytes` (#438)

* Deprecated `EdwardsPoint::hash_from_bytes` and renamed to
  `EdwardsPoint::nonspec_map_to_curve`

* Added KAT test vectors for `RistrettoPoint::from_uniform_bytes`

rozbb committed:
Updated CHANGELOG to reflect recent PRs

pinkforest committed:
Documentation migrate to docs.rs hosted

This change migrates all the documentation from dalek.rs
to docs.rs hosted and fixed the backend documentation
generation that was broken.

rozbb committed:
Fixed broken latex in parallel-formulas.md

rozbb committed:
Added cfg_attr everywhere possible, and simplified cfg over std/alloc

rozbb committed:
Removed unnecessary unstable features

rozbb committed:
Fixed features to tell docs.rs to use

rozbb committed:
Shouldn't have removed stdsimd feature

pinkforest committed:
Fix curve_models link for parallel formulas doc

pinkforest committed:
Document backend mod as INTERNALS:

rozbb committed:
Cleanup: enabled doc_auto_cfg and doc_cfg_hide

rozbb committed:
Applied @pinkforest's patch to make `make doc` build on non-x86_64 arches

rozbb committed:
Fixed README image layout and gave the logo the alt text described in #89

cbeck88 committed:
Relax Rng trait bounds to allow `?Sized` Rngs (#394)

This allows the code to compile if you pass it `&mut dyn RngType`,
since trait objects are unsized.

See here for an example of caller code that is simplified by this
change:

https://github.com/mobilecoinfoundation/mobilecoin/pull/1977#discussion_r872906913

rozbb committed:
Merge pull request #439 from pinkforest/docs-cleanup

rozbb committed:
Merge pull request #440 from dalek-cryptography/docs-cleanup

pinkforest committed:
Fix all clippy warnings replay (#441)

Also fixes CI not running on all branches

Co-authored-by: Anthony Ramine <nox@nox.paris>

pinkforest committed:
Minor documentation fixes (#444)

* Docs unlink from dalek.rs

* Link katex assets to jsdelivr

Co-authored-by: Michael Rosenberg <michael@mrosenberg.pub>

pinkforest committed:
Add target u32/u64 backend override (#454)

As suggested in #453 it is sometimes feasible to
select the backend bits via an override.

This change provides `cfg(curve25519_dalek_bits)`
to override the bits used in serial or fiat target backend.

pinkforest committed:
Make `rand_core` optional (#447)

As proposed in #442 this makes `rand_core` an
optional feature that is not covered by the
SemVer public API stability guarantees.

Co-authored-by: Michael Rosenberg <michael@mrosenberg.pub>

pinkforest committed:
Make `digest` optional (#446)

As proposed in #442 this makes `digest` an
optional feature that is not covered by the
SemVer public API stability guarantees.

Co-authored-by: Michael Rosenberg <michael@mrosenberg.pub>

tarcieri committed:
Remove `std` feature (#459)

All of the existing usages of `std` can be replaced with `alloc`.

They are legacy usages from before when liballoc was stabilized.

rozbb committed:
Remove mocks (#460)

Gated random() construtors on cfg(test)

rozbb committed:
Made Scalar::bits return an iterator rather than an array (#451)

Addresses issue #448 that Scalar::bits may leave unzeroed bits on the stack

tarcieri committed:
Fix `--no-default-features` test warnings; consolidate CI jobs (#461)

Previously `cargo test --no-default-features` would succeed but with
warnings. This commit fixes all of those warnings and tests
`--no-default-features` in CI to ensure that in perpetuity.

rozbb committed:
Updates to README (#453)

* Restructure README and CHANGELOG
* Explain semver policy
* Specify feature flags and backends more explicitly
* Remove nightly from the CI bc that didn't belong there
* Add @pinkforest to thankyou list

Co-authored-by: pinkforest <36498018+pinkforest@users.noreply.github.com>

tarcieri committed:
Use `--cfg curve25519_dalek_backend` to select backend (#455)

Crate features are intended to be additive, whereas only 1-of-N possible
backends can be selected.

Features can also be activated by transitive dependencies, which leads
to a problem of different dependences selecting conflicting backends.
Using `--cfg` instead moves all backend selection control to the
toplevel executable.

This commit switches to the following RUSTFLAGS to enable backends:

- `--cfg curve25519_dalek_backend="fiat"`: uses `fiat-crypto`
- `--cfg curve25519_dalek_backend="simd"`: uses nightly-only SIMD

pinkforest committed:
Fixes curve25519_dalek_bits defaults for cross and wasm (#465)

build.rs was using cfg(target) but it has to evaluate this from env TARGET
as build.rs cfg(target) in build context is the builder host and not the target.

This change fixes curve25519_dalek_bits lottery to determine the correct
automatic curve25119_dalek_bits with the help of platforms crate.

As discussed in #456 this also prepares for well known defaults for wasm and
arm serial backend via cfg(curve25519_dalek_bits = "64")

If the wasm32 or armv7 are going to be u64 serial by default these will be
followed up on later.

tarcieri committed:
Use `CryptoRngCore` trait (#469)

This is a convenience/marker trait for types which impl `CryptoRng` +
`RngCore` which makes the type signatures a little more readable.

It was introduced in `rand_core` v0.6.4 (now pinned as the minimum
version)

pinkforest committed:
Document `curve25519_dalek_` `backend` and `bits` selection (#467)

Co-authored-by: Michael Rosenberg <michael@mrosenberg.pub>

tarcieri committed:
Use inherent constants for `ZERO`, `ONE`, and `MINUS_ONE` (#470)

For the field element types `FieldElement` and `Scalar`, use inherent
constants instead of (non-const) functions to return these constant
values.

It's likely the original functions predate support for inherent
constants, but now that they're available, they're a better fit for
these sort of constant values.

rozbb committed:
Updated CHANGELOG; added stable/beta instructions to README

rozbb committed:
Merge branch 'main' into release/4.0

rozbb committed:
README: Removed non-existent features

pinkforest reviewed (approved):
Just one entry needs to be removed:
- "Set well known defaults for wasm and arm"

str4d reviewed (commented):
A couple of notes (not sure if the intent is for them to be fixed as PRs to `release/4.0` before this PR is merged, or pushed to this PR directly).

str4d commented:
The other breaking change I think we should consider is to remove `vartime_double_scalar_mul_basepoint` and replace it with `vartime_check_double_scalar_mul_basepoint` (suggested in https://github.com/dalek-cryptography/curve25519-dalek/pull/323#issuecomment-623578421). This would make #323 a performance improvement. The main question is whether the new API should include a `mul_by_cofactor` or not; `vartime_double_scalar_mul_basepoint` is currently used for both strict and non-strict verification in `ed25519-dalek`, and replicating that in `vartime_check_double_scalar_mul_basepoint` would I think necessitate _both_ options (so maybe two APIs, or one API with a `bool` option).

tarcieri committed:
Change `Scalar::from_canonical_bytes` to return `CtOption` (#472)

This is helpful for implementing `ff::PrimeField::from_repr`.
Also changes `Scalar::is_canonical` to return `Choice`.

rozbb commented on CHANGELOG.md:
```diff
@@ -3,6 +3,39 @@
+* Update backend selection to be more automatic
+* Remove `std` feature flag
+* Remove `nightly` feature flag
+* Replace methods `Scalar::{zero, one}` with constants `Scalar::{ZERO, ONE}`
```
`MINUS_ONE` is not in the public API of `Scalar`. I had to recompile docs to check that. Also docs are broken, so will fix.

str4d replied:
It's hard to tell sometimes, due to the existing codebase using `pub fn` for methods on private types. A nice cleanup would be to make those all `pub(crate) fn` so that visibility is more obvious locally.

rozbb committed:
Add `rust-version` to cargo

Co-authored-by: str4d <thestr4d@gmail.com>

rozbb committed:
Added latest changelog entries to readme

rozbb commented:
@str4d I think deprecating `vartime_double_scalar_mul_basepoint` should be done a little later on. I'm not convinced nobody is using it. But yes I agree that we should merge the lattice-trick lineq check. Possibly for 4.1?

rozbb commented:
> Just one entry needs to be removed:
> 
>     * "Set well known defaults for wasm and arm"

What do you mean? Is that a commit?

pinkforest commented:
> What do you mean? Is that a commit?

a line in the CHANGELOG.md

pinkforest commented on CHANGELOG.md:
```diff
@@ -3,6 +3,41 @@
+
+#### Other changes
+
+* Set well known defaults for wasm and arm
```
@rozbb This. We didn't end up doing this.

rozbb committed:
Removed changelog line about wasm and arm defaults

str4d commented:
> I think deprecating `vartime_double_scalar_mul_basepoint` should be done a little later on. I'm not convinced nobody is using it.

I think the point made in #323 was not that `vartime_double_scalar_mul_basepoint` was unused, but that (almost) every use of it could be replaced by `vartime_check_double_scalar_mul_basepoint` (modulo the `mul_by_cofactor` question), and the remaining uses were better served by general MSM logic (rather than a specialised length-2 MSM).

> But yes I agree that we should merge the lattice-trick lineq check. Possibly for 4.1?

Sure, if the change is merged post-4.0 then it just becomes an API expansion (as I was trying to do in the 2.0 release cycle), and removal of `vartime_double_scalar_mul_basepoint` would need to be deferred to 5.0 at the earliest.

rozbb committed:
Fixed docs build (#475)

Also sets code font size in docs back to normal (no longer small)

rozbb committed:
Bumped prerelease version

rozbb commented:
> I think the point made in #323 was not that `vartime_double_scalar_mul_basepoint` was unused, but that (almost) every use of it could be replaced by `vartime_check_double_scalar_mul_basepoint`

Right. I'm not 100% certain yet that every use could be replaced by `vartime_check_double_scalar_mul_basepoint`. More pragmatically: it took nontrivial thought and effort to deprecate the much more straightforward `EdwardsPoint::hash_from_bytes`. I'm a bit hesitant to start the process of deprecating a different thing now, especially because we really need to get _something_ out the door.

pinkforest commented:
If it needs a deprecation we could always feature-gate it for 4.0.1 as an optional additive non-breaking feature and then make it default 5.x after battletestign in the wild if so desired - like we're doing for the wasm/armv7 `curve25519_dalek_bits` case.

I think for 4.0.x also we could have auto-detection for backends as an optional additive non-breaking feature `auto` and then make it default in 5.x.

e.g. then this would make:

4.0.0 was mostly paying down the maintenance techdebt that is breaking people's builds atm and causing dependnecy duplication leading to long buildtimes :partying_face: 

5.0.0 could be a performance oriented release that contains a lot of well known better defaults for performance.

Also I've proposed some feature flags which we could include this type of work via earlier perhaps: 
- https://github.com/dalek-cryptography/curve25519-dalek/issues/468

pinkforest reviewed (approved):
LGTM :tada:

dignifiedquire commented:
I am so glad to see this happening 😍 thank you all!
